File: csv_editor.rb
# csv_editor.rb
require 'csv'

class CSVEditor
  def initialize(filename = nil)
    @filename = filename
    @headers = []
    @rows = []
    if filename && File.exist?(filename)
      load(filename)
    elsif filename
      create_empty
    else
      create_empty
    end
  end

  def load(filename)
    data = CSV.read(filename)
    return if data.empty?
    @headers = data[0].map(&:strip)
    @rows = data[1..-1].map { |row| row.map(&:strip) } if data.size > 1
  end

  def save(filename = nil)
    filename ||= @filename
    if filename.nil? || filename.empty?
      print "Enter filename to save: "
      filename = gets.chomp.strip
      filename += '.csv' unless filename.end_with?('.csv')
    end
    CSV.open(filename, 'w') do |csv|
      csv << @headers
      @rows.each { |row| csv << row }
    end
    puts "Saved."
  end

  def create_empty
    puts "Creating new CSV file."
    print "Enter column headers (comma-separated): "
    input = gets.chomp.strip
    @headers = input.empty? ? ['Column1', 'Column2', 'Column3'] : input.split(',').map(&:strip)
    @rows = []
    save(@filename || 'new.csv')
  end

  def display
    if @headers.empty?
      puts "No data loaded."
      return
    end
    col_widths = @headers.map(&:length)
    @rows.each do |row|
      row.each_with_index do |cell, i|
        col_widths[i] = cell.length if i < col_widths.length && cell.length > col_widths[i]
      end
    end
    sep = ' | '
    puts
    puts @headers.each_with_index.map { |h, i| h.ljust(col_widths[i]) }.join(sep)
    puts '-' * col_widths.sum + ('-' * (col_widths.length - 1) * sep.length)
    @rows.each do |row|
      cells = @headers.each_with_index.map { |_, i| (row[i] || '').ljust(col_widths[i]) }
      puts cells.join(sep)
    end
  end

  def add_row
    if @headers.empty?
      puts "No headers defined. Please add columns first."
      return
    end
    row = @headers.map do |h|
      print "Enter value for '#{h}': "
      gets.chomp.strip
    end
    @rows << row
    puts "Row added."
  end

  def edit_cell
    display
    print "Enter row index (1-based): "
    row_idx = gets.chomp.to_i - 1
    if row_idx < 0 || row_idx >= @rows.size
      puts "Invalid row index."
      return
    end
    puts "Columns: #{@headers.join(', ')}"
    print "Enter column name: "
    col_name = gets.chomp.strip
    col_idx = @headers.index(col_name)
    if col_idx.nil?
      puts "Column not found."
      return
    end
    print "Enter new value for '#{col_name}': "
    new_val = gets.chomp.strip
    @rows[row_idx][col_idx] = new_val
    puts "Cell updated."
  end

  def delete_row
    display
    print "Enter row index to delete (1-based): "
    row_idx = gets.chomp.to_i - 1
    if row_idx < 0 || row_idx >= @rows.size
      puts "Invalid row index."
      return
    end
    @rows.delete_at(row_idx)
    puts "Row deleted."
  end

  def filter_rows
    if @rows.empty?
      puts "No rows to filter."
      return
    end
    puts "Columns: #{@headers.join(', ')}"
    print "Enter column name to filter: "
    col_name = gets.chomp.strip
    col_idx = @headers.index(col_name)
    if col_idx.nil?
      puts "Column not found."
      return
    end
    print "Enter value to match: "
    value = gets.chomp.strip
    print "Case-sensitive? (y/n): "
    case_sensitive = gets.chomp.strip.downcase == 'y'
    filtered = @rows.select do |row|
      col_idx < row.size &&
      (case_sensitive ? row[col_idx] == value : row[col_idx].downcase == value.downcase)
    end
    if filtered.empty?
      puts "No matching rows."
    else
      old_rows = @rows
      @rows = filtered
      display
      @rows = old_rows
    end
  end

  def sort_rows
    if @rows.empty?
      puts "No rows to sort."
      return
    end
    puts "Columns: #{@headers.join(', ')}"
    print "Enter column name to sort by: "
    col_name = gets.chomp.strip
    col_idx = @headers.index(col_name)
    if col_idx.nil?
      puts "Column not found."
      return
    end
    print "Sort descending? (y/n): "
    reverse = gets.chomp.strip.downcase == 'y'
    @rows.sort_by! { |row| col_idx < row.size ? row[col_idx] : '' }
    @rows.reverse! if reverse
    puts "Rows sorted."
  end

  def add_column
    print "Enter new column name: "
    col_name = gets.chomp.strip
    if col_name.empty?
      puts "Invalid name."
      return
    end
    if @headers.include?(col_name)
      puts "Column already exists."
      return
    end
    print "Enter default value for existing rows (leave empty for empty string): "
    default_val = gets.chomp.strip
    @headers << col_name
    @rows.each { |row| row << default_val }
    puts "Column added."
  end

  def remove_column
    if @headers.empty?
      puts "No columns to remove."
      return
    end
    puts "Columns: #{@headers.join(', ')}"
    print "Enter column name to remove: "
    col_name = gets.chomp.strip
    col_idx = @headers.index(col_name)
    if col_idx.nil?
      puts "Column not found."
      return
    end
    @headers.delete_at(col_idx)
    @rows.each { |row| row.delete_at(col_idx) if col_idx < row.size }
    puts "Column removed."
  end

  def summary
    puts "File: #{@filename || '(new)'}"
    puts "Rows: #{@rows.size}"
    puts "Columns: #{@headers.size}"
    puts "Headers: #{@headers.join(', ')}" unless @headers.empty?
  end

  def run
    puts "=== CSV Editor/Viewer ==="
    if @filename
      puts "Loaded: #{@filename} (#{@rows.size} rows, #{@headers.size} columns)"
    else
      puts "No file loaded."
    end
    loop do
      puts "\nMenu:"
      puts "1. View table"
      puts "2. Add row"
      puts "3. Edit cell"
      puts "4. Delete row"
      puts "5. Filter rows"
      puts "6. Sort rows"
      puts "7. Add column"
      puts "8. Remove column"
      puts "9. Export / Save"
      puts "10. Show summary"
      puts "11. Exit"
      print "Choose: "
      choice = gets.chomp.strip
      case choice
      when '1' then display
      when '2' then add_row
      when '3' then edit_cell
      when '4' then delete_row
      when '5' then filter_rows
      when '6' then sort_rows
      when '7' then add_column
      when '8' then remove_column
      when '9' then save
      when '10' then summary
      when '11'
        print "Save before exit? (y/n): "
        save if gets.chomp.strip.downcase == 'y'
        puts "Goodbye!"
        break
      else puts "Invalid choice."
      end
    end
  end
end

filename = ARGV.first
editor = CSVEditor.new(filename)
editor.run

File: CsvEditor.cs
// CsvEditor.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

class CsvEditor
{
    private string filename;
    private List<string> headers;
    private List<List<string>> rows;

    public CsvEditor(string filename)
    {
        this.filename = filename;
        headers = new List<string>();
        rows = new List<List<string>>();
        if (!string.IsNullOrEmpty(filename) && File.Exists(filename))
            Load(filename);
        else
            CreateEmpty();
    }

    private void Load(string filename)
    {
        var lines = File.ReadAllLines(filename, Encoding.UTF8);
        if (lines.Length > 0)
        {
            headers = lines[0].Split(',').Select(h => h.Trim()).ToList();
            for (int i = 1; i < lines.Length; i++)
            {
                if (!string.IsNullOrWhiteSpace(lines[i]))
                    rows.Add(lines[i].Split(',').Select(c => c.Trim()).ToList());
            }
        }
    }

    private void Save(string filename = null)
    {
        if (string.IsNullOrEmpty(filename)) filename = this.filename;
        if (string.IsNullOrEmpty(filename))
        {
            Console.Write("Enter filename to save: ");
            filename = Console.ReadLine().Trim();
            if (!filename.EndsWith(".csv")) filename += ".csv";
        }
        var lines = new List<string>();
        lines.Add(string.Join(",", headers));
        foreach (var row in rows)
            lines.Add(string.Join(",", row));
        File.WriteAllLines(filename, lines, Encoding.UTF8);
        Console.WriteLine("Saved.");
    }

    private void CreateEmpty()
    {
        Console.WriteLine("Creating new CSV file.");
        Console.Write("Enter column headers (comma-separated): ");
        var input = Console.ReadLine()?.Trim() ?? "";
        if (string.IsNullOrEmpty(input))
            headers = new List<string> { "Column1", "Column2", "Column3" };
        else
            headers = input.Split(',').Select(h => h.Trim()).Where(h => !string.IsNullOrEmpty(h)).ToList();
        rows = new List<List<string>>();
        Save();
    }

    private void Display()
    {
        if (headers.Count == 0)
        {
            Console.WriteLine("No data loaded.");
            return;
        }
        var colWidths = headers.Select(h => h.Length).ToArray();
        foreach (var row in rows)
        {
            for (int i = 0; i < row.Count && i < colWidths.Length; i++)
                if (row[i].Length > colWidths[i]) colWidths[i] = row[i].Length;
        }
        var sep = " | ";
        Console.WriteLine();
        var headerLine = string.Join(sep, headers.Select((h, i) => h.PadRight(colWidths[i])));
        Console.WriteLine(headerLine);
        Console.WriteLine(new string('-', headerLine.Length));
        foreach (var row in rows)
        {
            var cells = row.Select((c, i) => i < colWidths.Length ? c.PadRight(colWidths[i]) : "").ToArray();
            Console.WriteLine(string.Join(sep, cells));
        }
    }

    private void AddRow()
    {
        if (headers.Count == 0)
        {
            Console.WriteLine("No headers defined. Please add columns first.");
            return;
        }
        var row = new List<string>();
        foreach (var h in headers)
        {
            Console.Write($"Enter value for '{h}': ");
            row.Add(Console.ReadLine()?.Trim() ?? "");
        }
        rows.Add(row);
        Console.WriteLine("Row added.");
    }

    private void EditCell()
    {
        Display();
        Console.Write("Enter row index (1-based): ");
        if (!int.TryParse(Console.ReadLine(), out int rowIdx) || rowIdx < 1 || rowIdx > rows.Count)
        {
            Console.WriteLine("Invalid row index.");
            return;
        }
        rowIdx--;
        Console.WriteLine("Columns: " + string.Join(", ", headers));
        Console.Write("Enter column name: ");
        var colName = Console.ReadLine()?.Trim();
        var colIdx = headers.IndexOf(colName);
        if (colIdx == -1)
        {
            Console.WriteLine("Column not found.");
            return;
        }
        Console.Write($"Enter new value for '{colName}': ");
        var newVal = Console.ReadLine()?.Trim() ?? "";
        rows[rowIdx][colIdx] = newVal;
        Console.WriteLine("Cell updated.");
    }

    private void DeleteRow()
    {
        Display();
        Console.Write("Enter row index to delete (1-based): ");
        if (!int.TryParse(Console.ReadLine(), out int rowIdx) || rowIdx < 1 || rowIdx > rows.Count)
        {
            Console.WriteLine("Invalid row index.");
            return;
        }
        rows.RemoveAt(rowIdx - 1);
        Console.WriteLine("Row deleted.");
    }

    private void FilterRows()
    {
        if (rows.Count == 0)
        {
            Console.WriteLine("No rows to filter.");
            return;
        }
        Console.WriteLine("Columns: " + string.Join(", ", headers));
        Console.Write("Enter column name to filter: ");
        var colName = Console.ReadLine()?.Trim();
        var colIdx = headers.IndexOf(colName);
        if (colIdx == -1)
        {
            Console.WriteLine("Column not found.");
            return;
        }
        Console.Write("Enter value to match: ");
        var value = Console.ReadLine()?.Trim() ?? "";
        Console.Write("Case-sensitive? (y/n): ");
        var caseSensitive = Console.ReadLine()?.Trim().ToLower() == "y";
        var filtered = rows.Where(row => colIdx < row.Count &&
            (caseSensitive ? row[colIdx] == value : row[colIdx].ToLower() == value.ToLower())).ToList();
        if (filtered.Count == 0)
            Console.WriteLine("No matching rows.");
        else
        {
            var oldRows = rows;
            rows = filtered;
            Display();
            rows = oldRows;
        }
    }

    private void SortRows()
    {
        if (rows.Count == 0)
        {
            Console.WriteLine("No rows to sort.");
            return;
        }
        Console.WriteLine("Columns: " + string.Join(", ", headers));
        Console.Write("Enter column name to sort by: ");
        var colName = Console.ReadLine()?.Trim();
        var colIdx = headers.IndexOf(colName);
        if (colIdx == -1)
        {
            Console.WriteLine("Column not found.");
            return;
        }
        Console.Write("Sort descending? (y/n): ");
        var reverse = Console.ReadLine()?.Trim().ToLower() == "y";
        if (reverse)
            rows = rows.OrderByDescending(row => colIdx < row.Count ? row[colIdx] : "").ToList();
        else
            rows = rows.OrderBy(row => colIdx < row.Count ? row[colIdx] : "").ToList();
        Console.WriteLine("Rows sorted.");
    }

    private void AddColumn()
    {
        Console.Write("Enter new column name: ");
        var colName = Console.ReadLine()?.Trim();
        if (string.IsNullOrEmpty(colName))
        {
            Console.WriteLine("Invalid name.");
            return;
        }
        if (headers.Contains(colName))
        {
            Console.WriteLine("Column already exists.");
            return;
        }
        Console.Write("Enter default value for existing rows (leave empty for empty string): ");
        var defaultVal = Console.ReadLine()?.Trim() ?? "";
        headers.Add(colName);
        foreach (var row in rows)
            row.Add(defaultVal);
        Console.WriteLine("Column added.");
    }

    private void RemoveColumn()
    {
        if (headers.Count == 0)
        {
            Console.WriteLine("No columns to remove.");
            return;
        }
        Console.WriteLine("Columns: " + string.Join(", ", headers));
        Console.Write("Enter column name to remove: ");
        var colName = Console.ReadLine()?.Trim();
        var colIdx = headers.IndexOf(colName);
        if (colIdx == -1)
        {
            Console.WriteLine("Column not found.");
            return;
        }
        headers.RemoveAt(colIdx);
        foreach (var row in rows)
            if (colIdx < row.Count) row.RemoveAt(colIdx);
        Console.WriteLine("Column removed.");
    }

    private void Summary()
    {
        Console.WriteLine($"File: {filename ?? "(new)"}");
        Console.WriteLine($"Rows: {rows.Count}");
        Console.WriteLine($"Columns: {headers.Count}");
        if (headers.Count > 0)
            Console.WriteLine("Headers: " + string.Join(", ", headers));
    }

    public void Run()
    {
        Console.WriteLine("=== CSV Editor/Viewer ===");
        if (!string.IsNullOrEmpty(filename))
            Console.WriteLine($"Loaded: {filename} ({rows.Count} rows, {headers.Count} columns)");
        while (true)
        {
            Console.WriteLine("\nMenu:");
            Console.WriteLine("1. View table");
            Console.WriteLine("2. Add row");
            Console.WriteLine("3. Edit cell");
            Console.WriteLine("4. Delete row");
            Console.WriteLine("5. Filter rows");
            Console.WriteLine("6. Sort rows");
            Console.WriteLine("7. Add column");
            Console.WriteLine("8. Remove column");
            Console.WriteLine("9. Export / Save");
            Console.WriteLine("10. Show summary");
            Console.WriteLine("11. Exit");
            Console.Write("Choose: ");
            var choice = Console.ReadLine()?.Trim();
            switch (choice)
            {
                case "1": Display(); break;
                case "2": AddRow(); break;
                case "3": EditCell(); break;
                case "4": DeleteRow(); break;
                case "5": FilterRows(); break;
                case "6": SortRows(); break;
                case "7": AddColumn(); break;
                case "8": RemoveColumn(); break;
                case "9": Save(); break;
                case "10": Summary(); break;
                case "11":
                    Console.Write("Save before exit? (y/n): ");
                    if (Console.ReadLine()?.Trim().ToLower() == "y") Save();
                    Console.WriteLine("Goodbye!");
                    return;
                default: Console.WriteLine("Invalid choice."); break;
            }
        }
    }

    static void Main(string[] args)
    {
        var filename = args.Length > 0 ? args[0] : null;
        var editor = new CsvEditor(filename);
        editor.Run();
    }
}
